File: source/clock.cpp
#include <chrono>

#include "clock.hpp"
#include "dcpu.hpp"

namespace dcpu {

  void Clock::Execute() {
    if (interval) {
      auto now = std::chrono::high_resolution_clock::now();
      if (last_tick + std::chrono::duration<double>(interval / kClockFrequency) <= now) {
        last_tick = now;
        ticks += 1;
        if (message) {
          dcpu->Interrupt(message);
        }
      }
    }
  }

  void Clock::Connect(Dcpu *dcpu) {
    this->dcpu = dcpu;
  }

  void Clock::HandleHardwareInterrupt() {
    if (dcpu) {
      switch (dcpu->register_a) {
        case 0: {
          SetInterval(dcpu->register_b);
          break;
        }
        case 1: {
          dcpu->register_c = GetElapsedTicks();
          break;
        }
        case 2: {
          SetSoftwareInterrupt(dcpu->register_b);
          break;
        }
        default: break;
      }
    }
  }

  void Clock::SetSoftwareInterrupt(Word message) {
    this->message = message;
  }

  void Clock::SetInterval(Word interval) {
    this->interval = interval;
    ticks = 0;
    last_tick = std::chrono::high_resolution_clock::now();
  }

  Word Clock::GetElapsedTicks() {
    return ticks;
  }

}  // namespace dcpu

File: clock/clock.go
package clock

import (
	"time"

	"github.com/robertsdionne/dcpu"
)

// Clock implements Generic Clock (compatible).
type Clock struct {
	Interval uint16
	Message  uint16
	Ticks    uint16

	lastTick time.Time
}

const (
	duration  = time.Second / frequency
	frequency = 60
	id        = 0x12d0b402
	version   = 0x0001
)

const (
	setInterval = iota
	getTicks
	setInterruptMessage
)

// Execute runs the clock.
func (c *Clock) Execute(dcpu *dcpu.DCPU) {
	if c.Interval > 0 {
		now := time.Now()

		if now.Sub(c.lastTick) > time.Duration(c.Interval)*duration {
			c.lastTick = now
			c.Ticks++
			if dcpu != nil && c.Message > 0 {
				dcpu.Interrupt(c.Message)
			}
		}
	}
}

// GetID returns the Clock id.
func (c *Clock) GetID() uint32 {
	return id
}

// GetManufacturerID returns the Clock manufacturer id.
func (c *Clock) GetManufacturerID() uint32 {
	return 0
}

// GetVersion returns the Clock version.
func (c *Clock) GetVersion() uint16 {
	return version
}

// HandleHardwareInterrupt handles messages from the DCPU.
func (c *Clock) HandleHardwareInterrupt(dcpu *dcpu.DCPU) {
	switch dcpu.RegisterA {
	case setInterval:
		c.Interval = dcpu.RegisterB
		c.Ticks = 0

	case getTicks:
		dcpu.RegisterC = c.Ticks

	case setInterruptMessage:
		c.Message = dcpu.RegisterB
	}
}
